### PR TITLE
CIRC-1357 Delete notice if failure or exception occurred

### PR DESCRIPTION
**Pull request by Mykyta-Varenyk**
Resolves: https://issues.folio.org/browse/CIRC-1357. 

## Purpose
If any errors during notice processing occur, this specific notice is just being skipped. Invalid notices can block the valid ones from being processed because of the notice processing limit.
## Approach
Notices are being deleted when any type of error occurs.

### CODE REVIEW

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java [hunk 1]
```diff
@@ -384,11 +403,28 @@ void testNoticeIsNotSentOrDeletedWhenPatronNoticeRequestFails() {
...
     checkSentNotices();
 
-    verifyNumberOfScheduledNotices(1);
+    verifyNumberOfScheduledNotices(0);
     verifyNumberOfPublishedEvents(NOTICE, 0);
     verifyNumberOfPublishedEvents(NOTICE_ERROR, 1);
   }
 
+  @Test
+  void testErrorEventsAreSentWhenUpdateNoticeRequestFails() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java:412]: Test that
- notice is not sent
- event is not published
- notice is deleted

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java [hunk 2]
```diff
@@ -384,11 +403,28 @@ void testNoticeIsNotSentOrDeletedWhenPatronNoticeRequestFails() {
...
+    generateLoanAndScheduledNotices();
+
+    createNoticesOverTime(dueDate.minusMinutes(1)::minusHours, 2).forEach(
+      scheduledNoticesClient::create);
+
+    FakeStorageModule.setFailDeleteWithBadRequest(true);
+    FakeStorageModule.setFailUpdateWithBadRequest(true);
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java:419]: Make flags non-static

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java [hunk 3]
```diff
@@ -419,6 +455,43 @@ void testNoticesForNonExistentLoansDoNotBlockTheQueue() {
...
+
+    scheduledNoticeProcessingClient.runLoanNoticesProcessing(dueDate.minusSeconds(1));
+
+    UUID expectedSentTemplateId1 = UUID.fromString(
+      notices.get(1).getJsonObject("noticeConfig").getString("templateId"));
+
+    UUID expectedSentTemplateId2 = UUID.fromString(
+      notices.get(3).getJsonObject("noticeConfig").getString("templateId"));
+
+    checkSentNoticesForSpecifiedItemAndUserAndLoan(basedUponNod, jessica, jessicaNodLoan,
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java:487]:
```suggestion
    checkSentNotices(basedUponNod, jessica, jessicaNodLoan,
```

File: src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java [hunk 4]
```diff
@@ -492,6 +492,21 @@ void scheduledNoticesShouldNotBeSentWhenRequestWasNotFound() {
     verifyNumberOfPublishedEvents(NOTICE_ERROR, 1);
   }
 
+  @Test
+  void scheduledNoticesShouldNotBeSentWhenRequestIdWasNull() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java:496]:
```suggestion
  void scheduledNoticesShouldNotBeSentWhenRequestIdIsNull() {
```

File: src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java [hunk 5]
```diff
@@ -492,6 +492,21 @@ void scheduledNoticesShouldNotBeSentWhenRequestWasNotFound() {
...
+    prepareNotice();
+
+    JsonObject entries = scheduledNoticesClient.getAll().get(0);
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java:499]:
```suggestion
    JsonObject notice= scheduledNoticesClient.getAll().get(0);
```

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java [hunk 6]
```diff
@@ -306,6 +309,22 @@ void testNoticeIsDeletedIfItHasNoLoanId() {
     verifyNumberOfPublishedEvents(NOTICE_ERROR, 1);
   }
 
+  @Test
+  void testNoticeIsDeletedIfPatronGroupIsNull() {
+    generateLoanAndScheduledNotices();
+
+    JsonObject brokenNotice = createNoticesOverTime(dueDate.minusMinutes(1)::minusHours,
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java]: Done

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java (hunk 1 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java]: Test deleted, as we cannot access fake storage instances

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java (hunk 2 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java]: Deleted, as we cannot access fake storage instances

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java (hunk 3 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java]: Done

File: src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java (hunk 4 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java]: Done

File: src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java (hunk 5 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/requests/RequestScheduledNoticesProcessingTests.java]: Done

File: src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeHandler.java [hunk 7]
```diff
@@ -201,16 +201,12 @@ protected boolean noticeShouldNotBeSent(ScheduledNoticeContext context) {
     if (result.succeeded()) {
       log.info("Finished processing scheduled notice {}", notice.getId());
       return completedFuture(result);
-    }
-
-    HttpFailure failure = result.cause();
-    log.error("Processing scheduled notice {} failed: {}", notice.getId(), failure);
+    } else {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeHandler.java:204]: No need for `else`, there is a `return` in `if` above

File: src/test/java/api/loans/DueDateNotRealTimeScheduledNoticesProcessingTests.java [hunk 8]
```diff
@@ -396,20 +401,23 @@ void missingReferencedEntitiesDoNotBlockProcessing() {
...
-    checkOutFixture.checkOutByBarcode(times, steve, loanDate.plusHours(4));
-    IndividualResource uprootedToSteve = checkOutFixture.checkOutByBarcode(uprooted, steve, loanDate.plusHours(5));
-    checkOutFixture.checkOutByBarcode(dunkirk, jessica, loanDate.plusHours(6));
+    checkOutFixture.checkOutByBarcode(lotr, charlotte, loanDate.plusHours(3));
+    IndividualResource planetToJames = checkOutFixture.checkOutByBarcode(planet, james, loanDate.plusHours(4));
+    checkOutFixture.checkOutByBarcode(times, steve, loanDate.plusHours(5));
+    IndividualResource uprootedToSteve = checkOutFixture.checkOutByBarcode(uprooted, steve, loanDate.plusHours(6));
+    checkOutFixture.checkOutByBarcode(dunkirk, jessica, loanDate.plusHours(7));
 
+    usersClient.replace(charlotte.getId(), new UserBuilder().withPatronGroupId(null));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateNotRealTimeScheduledNoticesProcessingTests.java:415]: The test's name is `missingReferencedEntities`, and here you are deleting the reference, not the entity

File: src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java [hunk 9]
```diff
@@ -419,6 +437,44 @@ void testNoticesForNonExistentLoansDoNotBlockTheQueue() {
...
+    int expectedNumberOfUnprocessedNotices = 0;
+
+    List<JsonObject> notices = createNoticesOverTime(dueDate.minusMinutes(1)::minusHours, 4);
+
+    var basedUponNod = itemsFixture.basedUponNod();
+    var jessica = usersFixture.jessica();
+
+    var jessicaNodLoan = checkOutFixture.checkOutByBarcode(basedUponNod, jessica);
+
+    usersClient.replace(borrower.getId(), new UserBuilder().withPatronGroupId(null));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java:453]: Again, this is not what the test name implies. Please rename the test OR delete the patron group OR change `patronGroupId` to a non-existent one.

Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateScheduledNoticesProcessingTests.java:453]: Test renamed

File: src/test/java/api/loans/DueDateNotRealTimeScheduledNoticesProcessingTests.java (hunk 8 above)
Comment (Mykyta-Varenyk) [reply src/test/java/api/loans/DueDateNotRealTimeScheduledNoticesProcessingTests.java]: New test with appropriate name has been created

File: src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeHandler.java (hunk 7 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/circulation/domain/notice/schedule/ScheduledNoticeHandler.java]: Else removed

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**